multilinear-util/src/point.rs

```diff
@@ -0,0 +1,655 @@
+use core::ops::{Index, Range};
+
+use p3_field::Field;
+use rand::Rng;
+use rand::distr::{Distribution, StandardUniform};
+
+/// A point `(x_1, ..., x_n)` in `F^n` for some field `F`.
+#[derive(Default, Debug, Clone, PartialEq, Eq)]
+pub struct MultilinearPoint<F>(pub Vec<F>);
+
+impl<F> MultilinearPoint<F>
+where
+    F: Field,
+{
+    /// Returns the number of variables (dimension `n`).
+    #[inline]
+    #[must_use]
+    pub const fn num_variables(&self) -> usize {
+        self.0.len()
+    }
+
+    /// Return a reference to the slice of field elements
+    /// defining the point.
+    #[inline]
+    #[must_use]
+    pub fn as_slice(&self) -> &[F] {
+        &self.0
+    }
+
+    /// Return an iterator over the field elements making up the point.
+    #[inline]
+    pub fn iter(&self) -> core::slice::Iter<'_, F> {
+        self.0.iter()
+    }
+
+    /// Return a sub-point over the specified range of variables.
+    #[inline]
+    #[must_use]
+    pub fn get_subpoint_over_range(&self, idx: Range<usize>) -> Self {
+        Self(self.0[idx].to_vec())
+    }
+
+    /// Return a reference to the last variable in the point, if it exists.
+    ///
+    /// Returns None if the point is empty.
+    #[inline]
+    #[must_use]
+    pub fn last_variable(&self) -> Option<&F> {
+        self.0.last()
+    }
+
+    /// Converts a univariate evaluation point into a multilinear one.
+    ///
+    /// Uses the bijection:
+    /// ```ignore
+    /// f(x_1, ..., x_n) <-> g(y) := f(y^(2^(n-1)), ..., y^4, y^2, y)
+    /// ```
+    /// Meaning:
+    /// ```ignore
+    /// x_1^i_1 * ... * x_n^i_n <-> y^i
+    /// ```
+    /// where `(i_1, ..., i_n)` is the **big-endian** binary decomposition of `i`.
+    ///
+    /// Reversing the order ensures the **big-endian** convention.
+    pub fn expand_from_univariate(point: F, num_variables: usize) -> Self {
+        let mut cur = point;
+        let mut res = (0..num_variables)
+            .map(|_| {
+                let value_to_return = cur;
+                // Compute y^(2^k) at each step
+                cur = cur.square();
+                value_to_return
+            })
+            .collect::<Vec<F>>();
+
+        res.reverse();
+        Self(res)
+    }
+
+    /// Computes `eq(c, p)`, where `p` is another `MultilinearPoint`.
+    ///
+    /// The **equality polynomial** for two vectors is:
+    /// ```ignore
+    /// eq(s1, s2) = ∏ (s1_i * s2_i + (1 - s1_i) * (1 - s2_i))
+    /// ```
+    #[must_use]
+    pub fn eq_poly_outside(&self, point: &Self) -> F {
```

Comment:
Don't need the "outside" bit anymore as there isn't another eq_poly function for inside the binary hypercube.
```suggestion
    pub fn eq_poly(&self, point: &Self) -> F {
```